File: Graphs/839. Similar String Groups.py
class Solution:
    def numSimilarGroups(self, strings: List[str]) -> int:
        # for every node, connect another node such that 
        # t a r s  
        # r a t s 
        # a r t s 

        # ==== 1 group ====== 
        # s t a r 
        # 300 x 300 = 9 x 10 * 4 
        # 9 x 10**4 x 300 = 12 * 10**6 = 1.2 * 10**7
        # so for every string, try to connect a string to a group. 
        # so how do I see if there is a group? 
        if len(strings) == 1:
            return 1
        parents = [i for i in range(len(strings))]
        rank = [1 for i in range(len(strings))]
        res = len(strings)
        n = len(strings)
        def find(index):
            # here we find the parent of the currentString 
            current = index 
            while current != parents[current]:
                current = parents[current]
            return current

        def union(par1, par2):
            p1, p2 = find(par1), find(par2)
            if p1 == p2:
                return 0 
            
            # join the 2 strings here 
            if rank[p1] > rank[p2]:
                rank[p1] += rank[p2]
                rank[p2] = rank[p1]
                parents[p2] = parents[p1]
            else:
                rank[p2] += rank[p1]
                rank[p1] = rank[p2]
                parents[p1] = parents[p2]

            return -1         




        for i in range(n):
            for j in range(i + 1, n):
                string1 = strings[i]
                string2 = strings[j]

                if parents[i] == parents[j] :
                    continue
                    
                elif group(string1, string2):
                    res += union(i, j)

        return res
        
def group(string1, string2):
    diff = 0 
    for i in range(len(string1)):
        if string1[i] != string2[i]:
            diff += 1   
    return diff == 2 or diff == 0 

File: strings/839. Similar String Groups.cpp
class Solution {

public:
    int numSimilarGroups(vector<string>& strs) {
        int n = strs.size();
        vector<int> parents(n, 1);

        vector<int> ranks(n, 1);
        for (int i = 0; i < n; i++) {
            parents[i] = i;
        }
        int res = n;

        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (same(i, j, strs)) {
                    res += uf(i, j, parents, ranks);
                }


            }
        }

        return res;

    }

    bool same(int i, int j, vector<string>& strs) {
        int diff = 0;
        for (int z = 0; z < strs[i].length(); z++) {
            if (strs[i][z] != strs[j][z]) {
                diff++;
            }
        }

        if (diff == 2 || diff == 0) return true;

        return false;
    }


    int find(int x, vector<int>& parents) {
        while (x != parents[x]) {
            x = parents[x];
        }
        return x;
    }

    int uf(int i, int j, vector<int>& parents, vector<int>& ranks) {
        int par1 = find(i, parents);
        int par2 = find(j, parents);
        if (par1 == par2) {
            return 0;
        }

        if (ranks[par1] > ranks[par2]) {
            parents[par2] = par1;
            ranks[par1] += ranks[par2];
            ranks[par2] = ranks[par1];
        } else {
            parents[par1] = par2;
            ranks[par2] += ranks[par1];
            ranks[par1] = ranks[par2];
        }

        return -1;





    }




};